core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,152 @@
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+
+    // Pick modifications from the snapshot
+    Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    boolean isWapWorkflow =  wapId != null && !"".equals(wapId);
+    if (isWapWorkflow) {
+      if (base.isWapIdPublished(wapId)) {
+        throw new DuplicateWAPCommitException(wapId);
+      }
+    }
+    // only append operations are currently supported
+    if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+      throw new UnsupportedOperationException("Can cherry pick only append operations");
+    }
+    // this is to ensure we add files only once for each targetSnapshotId, to protect from
+    // duplicate additions if commit retries the same cherrypick operation on failure.
+    if (!snapshotsAlreadyCherrypicked.contains(targetSnapshotId)) {
+      for (DataFile addedFile : cherryPickSnapshot.addedFiles()) {
+        add(addedFile);
+      }
+      // this property is set on target snapshot that will get published
+      if (isWapWorkflow) {
+        set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+      }
+      snapshotsAlreadyCherrypicked.add(targetSnapshotId);
+    }
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots setCurrentSnapshot(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.ROLLBACK;
+    this.targetSnapshotId = snapshotId;
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackToTime(long timestampMillis) {
+    operation = SnapshotManagerOperation.ROLLBACK;
+    // find the latest snapshot by timestamp older than timestampMillis
+    Snapshot snapshot = SnapshotUtil.findLatestSnapshotOlderThan(base, timestampMillis);
+    Preconditions.checkArgument(snapshot != null,
+        "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+    this.targetSnapshotId = snapshot.snapshotId();
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackTo(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    ValidationException.check(SnapshotUtil.isCurrentAncestor(base, snapshotId),
+        "Not a valid snapshot to rollback to. Cannot rollback to a snapshot " +
+            "that's not an ancestor of the current snapshot.");
+    return setCurrentSnapshot(snapshotId);
+  }
+
+  @Override
+  public Snapshot apply() {
```

Comment:
Validations that can be violated on retry need to be done here instead of in the configuration methods. For example, say I have two concurrent processes picking the same WAP ID. One of them will win and pick the commit as the current snapshot. When the second one retries, it will have already validated that the WAP ID was not in the current state's ancestors and will go ahead. To fix this problem, the WAP check needs to be done here.

I don't think that the ancestor check needs to be done here, though. If concurrent rollbacks conflict, then I think it is okay if the last one wins.

Reply:
you'r right. thanks for catching this.